fem/nonlinearform_ext.hpp

```diff
@@ -17,28 +17,75 @@
 
 namespace mfem
 {
+
 class NonlinearForm;
+class NonlinearFormIntegrator;
 
+/** @brief Class extending the NonlinearForm class to support the different
+    AssemblyLevel%s. */
 class NonlinearFormExtension : public Operator
 {
 protected:
-   NonlinearForm *n; ///< Not owned
+   const NonlinearForm *nlf; ///< Not owned
+
 public:
-   NonlinearFormExtension(NonlinearForm *form);
-   virtual void AssemblePA() = 0;
+   NonlinearFormExtension(const NonlinearForm*);
+
+   /// Assemble at the AssemblyLevel of the subclass.
+   virtual void Assemble() = 0;
+   /// Assemble gradient data at the AssemblyLevel of the subclass, for the
+   /// state @a x, which is assumed to be a ldof Vector.
+   virtual void AssembleGradient(const Vector &x) = 0;
+
+   /// Assumes that @a x is a ldof Vector.
+   virtual Operator &GetGradient(const Vector &x) const = 0;
+
+   /// Assumes that @a x is a ldof Vector.
+   virtual double GetGridFunctionEnergy(const Vector &x) const = 0;
 };
 
 /// Data and methods for partially-assembled nonlinear forms
 class PANonlinearFormExtension : public NonlinearFormExtension
 {
+private:
+   class Gradient : public Operator
+   {
+   protected:
+      const Operator *elemR;
+      const FiniteElementSpace &fes;
+      const Array<NonlinearFormIntegrator*> &dnfi;
+      mutable Vector ge, xe, ye, ze;
```

Comment:
I still believe that two of these vectors could use the storage from the vectors `xe` and `ye` that are already created when `PANonlinearFormExtension` is made. However, you could do this in a separate PR as further optimizations to the class.